### PR TITLE
Make Assembly and extra element integer output action consistent

### LINKED ISSUE #27024: Assembly::extraElemID may get an extra element integer index out of bound
### Motivation

I am having a situation where extra element integers are added by user objects other than mesh generators. Unfortunately, `Assembly` constructor resizes `_extra_elem_ids` for storing extra element ids for the current element, which happens before extra element integers are added. When we add an `ExtraElementIDAux` for the late extra element integers in `ElementIDOutputAction`, we will pass an index out side of the bound of `_extra_elem_ids`. We did not hit valgrind errors I guess because of the vector capacity is greater than the vector size.

### Design

To allow adding extra element integers after mesh generation, we can resize `Assembly::_extra_elem_ids` with a larger size. Typically the number of extra element integers added by users are limited, thus this won't become a memory issue. We cannot resize `Assembly::_extra_elem_ids` because it may change the address of the data held in the vector making addresses returned by previous calls to `Assembly::extraElemID` invalid.

### Impact

Fix potential issues related with `ElementIDOutputAction`. Extra element integers added by user objects are typically used internally and are not to be referred with `Assembly::extraElemID`.

### PR DESCRIPTION
**Pull request by YaqiWang**
Closes #27024.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This will require a huge Griffin trivial regolding because we were not careful on these extra element integers.

**Comment by YaqiWang** [comment]
I will start preparing a Griffin patch once you agree on this change and will let you know when the patch is ready. Thank you.

File: framework/src/actions/ElementIDOutputAction.C [hunk 1]
```diff
@@ -55,7 +56,7 @@ ElementIDOutputAction::act()
         auto var_params = _factory.getValidParams("MooseVariableConstMonomial");
         auto kernel_params = _factory.getValidParams("ExtraElementIDAux");
         kernel_params.set<ExecFlagEnum>("execute_on") = EXEC_INITIAL;
-        for (unsigned int i = 0; i < _mesh->getMesh().n_elem_integers(); ++i)
+        for (unsigned int i = 0; i < _problem->assembly(0, 0).numExtraElemIntegers(); ++i)
```
Comment (lindsayad) [inline framework/src/actions/ElementIDOutputAction.C:59]: Is this change necessary? It seems like this PR could be logically split in two. The first would be the added assertion checking ensuring that someone doesn't get an invalid ID from `Assembly`. That's definitely a needed change.

I'm less sure about this one. The name of this class is not `ElementIDsHeldByAssemblyOutputAction`. Do you need the ids output here to be consistent with the ids help by `Assembly`? Now this output action is inconsistent with the extra elem ids held by the *Mesh*

Comment (lindsayad) [reply framework/src/actions/ElementIDOutputAction.C:59]: never mind. I see the issue since we go through assembly for the aux kernel

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
let me know when you have the Griffin patch ready

**Comment by YaqiWang** [comment]
Thanks @lindsayad. I guess I do not need to answer that comment. I will prepare Griffin patch now. It requires tons of trivial regoldings.

File: framework/src/actions/ElementIDOutputAction.C (hunk 1 above)
Comment (YaqiWang) [reply framework/src/actions/ElementIDOutputAction.C:59]: ExtraElementIDAux this action adds requires Assembly.

**Comment by YaqiWang** [comment]
@lindsayad patch in Griffin is ready.

**Comment by roystgnr** [comment]
Do users really save pointers to the references returned by `extraElemId*` here?

In the past I myself have long wavered between "setters and getters should be two separate functions because that's the only way to get real encapsulation" versus "just be lazy and return a reference"... but I didn't even think of "users will hang on to a pointer indefinitely" as a negative consequence of taking the terse option.  Is it unfair to expect a random C++ user to anticipate that, if they take `&foo[i]`, that's not guaranteed to be dereferenceable forever?

**Comment by roystgnr** [comment]
> Do users really save pointers to the references returned by extraElemId* here?

Thanks to Yaqi for clarifying on Teams:

Of course users save pointers to these references here, because we change the data at those pointers as we iterate from element to element during an assembly; it's the same design idiom as our pointers to shape function data.  But because objects typically only save those pointers once, during their construction, we can't even resize that vector in between one assembly and another, because we don't reconstruct e.g. a kernel in between assembly calls.

**Comment by YaqiWang** [comment]
@lindsayad Patch in Griffin is ready. Wanna merge this?

**Pull request merged**